app/logic/certification.py

```diff
@@ -1,14 +1,68 @@
 from peewee import JOIN, DoesNotExist, Case
-
+from flask import g
+from app.models.event import Event
+from app.models.term import Term
 from app.models.certification import Certification
 from app.models.certificationRequirement import CertificationRequirement
 from app.models.requirementMatch import RequirementMatch
 from app.models.eventParticipant import EventParticipant
+from app.models.user import User
+
+def termsAttended(certification=None, username=None):
+    '''
+    Retrieve terms attended by a user for certification and filter them based on frequency of a term
+    '''
+    attendedTerms = []
+    if username:
+        attendance = (RequirementMatch.select()
+                                      .join(EventParticipant, JOIN.LEFT_OUTER, on=(RequirementMatch.event == EventParticipant.event)) 
+                                      .where(RequirementMatch.requirement_id == certification)  
+                                      .where(EventParticipant.user == username))  
+    for termRecord in range(len(attendance)):
+        attendedTerms.append(attendance[termRecord].event.term.description)
+    return attendedTerms
+            
+def termsMissed(certification=None, username=None): 
+    '''
+    Calculate how many certification-eligible terms a student has missed based on their class level
+    and attendance record.
+
+    Logic:
+    - Each class level is expected to participate in 2 terms per year.
+    - If the user is currently in their final spring term (e.g., Spring of senior year), 
+      they are expected to have completed all terms: missedTerms = (level + 1) * 2.
+    - Otherwise, assume they’ve had one fewer term to attend: missedTerms = ((level + 1) * 2) - 1.
+    - If the user's classification is None, assume just 1 expected term.
+    - Subtract the number of terms the student has attended from the expected total to get the missed count.
+    '''
+    classLevel = ["Freshman", "Sophomore", "Junior", "Senior"]
+    currentTerm = g.current_term 
+    currentDescription = currentTerm.description
+    
+    # looking into a scenario where the current term is summer so that we can reassigned the current term variable to the next term     
+    if currentTerm.isSummer == True:
+        currentDescription = f'Fall {currentTerm.year}'
+        currentTerm = Term.select(Term).where(Term.description == currentDescription).get()
+    else:
+        currentDescription = currentTerm.description
+
+    for level in range(4):
+        user = User.select().where(User.username == username).get()
```

Comment:
We don't need to get the user every time in the loop, we can get it before the loop. In this particular case, we can get the user at the top of the function to validate the `username` parameter.